### PR TITLE
move runtime definition from cli-v2 to csdp-official

### PR DESCRIPTION
**Pull request by danielm-codefresh**

### CODE REVIEW

File: cmd/commands/runtime_install.go [hunk 1]
```diff
@@ -2131,5 +2129,5 @@ func getRuntimeDef(runtimeDef, versionStr string) string {
 	}
 
 	// specific version means the runtimeDef is the default value in cli-v2 repo
```
Comment (ghost) [inline cmd/commands/runtime_install.go]: maybe just fix the comment to mention the default is in csdp-official now

File: Makefile [hunk 2]
```diff
@@ -11,7 +11,7 @@ RUNTIME_DEF_URL="https://github.com/codefresh-io/cli-v2/releases/latest/download
 ADD_CLUSTER_DEF_URL="https://github.com/codefresh-io/csdp-official/add-cluster/kustomize"
 FALLBACK_ADD_CLUSTER_DEF_URL="https://github.com/codefresh-io/cli-v2/manifests/add-cluster/kustomize"
 
-DEV_RUNTIME_DEF_URL="manifests/runtime.yaml"
+DEV_RUNTIME_DEF_URL="<csdp-official-clone>/csdp/hybrid/runtime.yaml"
```
Comment (ghost) [inline Makefile]: how is that supposed to work? this is the value that will be injected to the store in `make local`

File: pkg/store/store.go [hunk 3]
```diff
@@ -31,11 +31,13 @@ var (
 	buildDate                = ""
 	gitCommit                = ""
 	SegmentWriteKey          = ""
-	maxDefVersion            = "2.1.1"
-	RuntimeDefURL            = "manifests/runtime.yaml"
+	maxDefVersion            = "2.1.2"
```
Comment (roi-codefresh) [inline pkg/store/store.go:35]: I don't think we need this anymore

File: pkg/store/store.go [hunk 4]
```diff
@@ -151,16 +155,33 @@ type Store struct {
 	IngressHost                       string
 	IscRuntimesDir                    string
 	DevMode                           bool
+	DefVersionComptability            map[string]string
```
Comment (roi-codefresh) [inline pkg/store/store.go]:
```suggestion
	DefVersionToLastCLIVersion            map[string]string
```

File: pkg/store/store.go [hunk 5]
```diff
@@ -151,16 +155,33 @@ type Store struct {
...
+	if versionStr == "" {
+		return runtimeDefURL
+	}
+
+	version := semver.MustParse(versionStr)
+	if version.Compare(s.LastRuntimeVersionInCLI) <= 0 {
+		runtimeDefURL = OldRuntimeDefURL
+	}
+
+	return runtimeDefURL
```
Comment (roi-codefresh) [inline pkg/store/store.go]: let's to where it is called

File: cmd/commands/runtime.go [hunk 6]
```diff
@@ -817,6 +817,8 @@ func NewRuntimeUpgradeCommand() *cobra.Command {
 				finalParameters["Version"] = versionStr
```
Comment (roi-codefresh) [inline cmd/commands/runtime.go]: remove `versionStr` var, looks dead

Comment (roi-codefresh) [reply cmd/commands/runtime.go]: replace with opts.versionStr

File: cmd/commands/runtime.go [hunk 7]
```diff
@@ -878,6 +880,14 @@ func runRuntimeUpgrade(ctx context.Context, opts *RuntimeUpgradeOptions) error {
 		return err
 	}
 
+	requiredCliVersionConstraint, _ := semver.NewConstraint(newRt.Spec.RequiredCLIVersion)
```
Comment (roi-codefresh) [inline cmd/commands/runtime.go]: explain why ignoring the err here, what are the possible flows here

File: cmd/commands/runtime.go [hunk 8]
```diff
@@ -878,6 +880,14 @@ func runRuntimeUpgrade(ctx context.Context, opts *RuntimeUpgradeOptions) error {
...
+	if requiredCliVersionConstraint != nil && !requiredCliVersionConstraint.Check(store.Get().Version.Version) {
+		err = fmt.Errorf("to upgrade to this version, please install cli version %s", requiredCliVersionConstraint.String())
```
Comment (roi-codefresh) [inline cmd/commands/runtime.go]:
```suggestion
		return fmt.Errorf("to upgrade to this version, please install cli version %s", requiredCliVersionConstraint.String())
```

File: cmd/commands/runtime_install.go [hunk 9]
```diff
@@ -222,6 +222,8 @@ func NewRuntimeInstallCommand() *cobra.Command {
 				return util.DecorateErrorWithDocsLink(fmt.Errorf("pre installation error: %w", err), store.Get().RequirementsLink)
 			}
 
+			installationOpts.runtimeDef = store.GetRuntimeDefURL(installationOpts.versionStr)
```
Comment (roi-codefresh) [inline cmd/commands/runtime_install.go]: need to check if this is assigned before overwriting

Comment (roi-codefresh) [reply cmd/commands/runtime_install.go]: for upgrade also

File: cmd/commands/runtime_install.go [hunk 10]
```diff
@@ -1143,6 +1146,14 @@ func preInstallationChecks(ctx context.Context, opts *RuntimeInstallOptions) (*r
...
 		err = fmt.Errorf("your cli version is out of date. please upgrade to the latest version before installing")
+	} else if rt.Spec.DefVersion.LessThan(store.Get().MaxDefVersion) {
+		val := store.Get().DefVersionComptability[rt.Spec.DefVersion.String()]
+		err = fmt.Errorf("to install this version, please downgrade your cli to version %s", val)
+	}
+
+	requiredCliVersionConstraint, _ := semver.NewConstraint(rt.Spec.RequiredCLIVersion)
+	if requiredCliVersionConstraint != nil && requiredCliVersionConstraint.Check(store.Get().Version.Version) {
+		err = fmt.Errorf("to install this version, please install cli version %s", requiredCliVersionConstraint.String())
 	}
```
Comment (roi-codefresh) [inline cmd/commands/runtime_install.go]: please extract this operation to runtime.go pkg

File: cmd/commands/runtime_install.go [hunk 11]
```diff
@@ -1143,6 +1146,14 @@ func preInstallationChecks(ctx context.Context, opts *RuntimeInstallOptions) (*r
 
 	if rt.Spec.DefVersion.GreaterThan(store.Get().MaxDefVersion) {
 		err = fmt.Errorf("your cli version is out of date. please upgrade to the latest version before installing")
+	} else if rt.Spec.DefVersion.LessThan(store.Get().MaxDefVersion) {
```
Comment (roi-codefresh) [inline cmd/commands/runtime_install.go]: If defVersion is nil on the runtime spec you should ignore it

File: cmd/commands/runtime_install.go [hunk 12]
```diff
@@ -2135,6 +2146,6 @@ func getRuntimeDef(runtimeDef, versionStr string) string {
 		return runtimeDef
 	}
 
-	// specific version means the runtimeDef is the default value in cli-v2 repo
-	return strings.Replace(runtimeDef, "/releases/latest/download", "/releases/download/v"+version.String(), 1)
+	// specific version means the runtimeDef is the default value in cli-v2/csdp-official repo
+	return runtimeDef + "?ref=v" + version.String()
```
Comment (roi-codefresh) [inline cmd/commands/runtime_install.go]: check if new or old runtime

File: pkg/store/store.go (hunk 3 above)
Comment (roi-codefresh) [inline pkg/store/store.go:35]: please do not touch this field it is deprecated, it's only here to allow to install runtimes with version < xyz

**Review by roi-codefresh — APPROVED** [review]

**Pull request merged**